File: SlackAPIsAngular/src/app/services/git-hub.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs/Observable';
import {GitHubUser} from '../models/gitHubUser';
// import {Response} from '@angular/http';
import 'rxjs/add/operator/map';


@Injectable()
export class GitHubService {
  constructor(private http: HttpClient) {
  }


  uri: string = 'http://localhost:8080/gitHub/users/';

  getUserByUserName(userName: string): Observable<any> {
    return this.http.get(this.uri + userName).map(user => {
        if(user != null){
          return new GitHubUser(user);
        }
        return null;
      }
    );
  }

}

File: SlackAPIsSpring/src/main/java/com/example/APIs/services/GitHubService.java
package com.example.APIs.services;

import com.example.APIs.models.GitHubUser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import java.util.HashMap;

@Service
public class GitHubService {

    Logger logger = LoggerFactory.getLogger(GitHubService.class);

    private HashMap<String, GitHubUser> storedUsers = new HashMap<>();

    @Autowired
    RestTemplate restTemplate;

    public GitHubUser getUserByUserName(String userName){

        if (storedUsers.containsKey(userName)) {
            logger.info("Local call");
            return storedUsers.get(userName);
        }
        try {
            GitHubUser user = this.callAPI(userName);
            storedUsers.put(userName, user);
            return user;
        } catch (Exception e){
            logger.warn(e.getMessage());
            return null;
        }

    }

    private GitHubUser callAPI(String userName) throws Exception{
        logger.info("API call");
        String uri = "https://api.github.com/users/" + userName;
        return restTemplate.getForObject(uri, GitHubUser.class);
    }
}
